<!DOCTYPE html>
<!-- saved from url=(0075)http://genomicsclass.github.io/book/pages/inference_for_highthroughput.html -->
<html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <meta charset="utf-8">
    <title>Basic inference for high-throughput data</title>
    
    <meta name="author" content="Rafael Irizarry and Michael Love">

    <!-- Enable responsive viewport -->
    <meta name="viewport" content="width=device-width, initial-scale=1.0">

    <!-- Le HTML5 shim, for IE6-8 support of HTML elements -->
    <!--[if lt IE 9]>
      <script src="http://html5shim.googlecode.com/svn/trunk/html5.js"></script>
    <![endif]-->

    <!-- Le styles -->
    <link href="http://genomicsclass.github.io/book/assets/themes/twitter/bootstrap/css/bootstrap.2.2.2.min.css" rel="stylesheet">
    <link href="./Basic inference for high-throughput data_files/style.css" rel="stylesheet" type="text/css" media="all">
    <link href="./Basic inference for high-throughput data_files/kbroman.css" rel="stylesheet" type="text/css" media="all">
    <link href="./Basic inference for high-throughput data_files/knitr.css" rel="stylesheet" type="text/css" media="all">

    <!-- Le fav and touch icons -->
    <link href="http://genomicsclass.github.io/book/assets/themes/twitter/ico/favicon.ico" rel="icon" type="image/png">

    <!-- atom & rss feed -->
    <link href="http://genomicsclass.github.io/booknil" type="application/atom+xml" rel="alternate" title="Sitewide ATOM Feed">
    <link href="http://genomicsclass.github.io/booknil" type="application/rss+xml" rel="alternate" title="Sitewide RSS Feed">

    <!-- knitr javascript -->
    <script src="./Basic inference for high-throughput data_files/knitr.js" type="text/javascript"></script>

    <!-- mathjax javascript -->
    <script type="text/x-mathjax-config;executed=true">
      MathJax.Hub.Config({
      tex2jax: {inlineMath: [['$','$'], ['\\(','\\)']]}
      });
    </script>

    <script src="./Basic inference for high-throughput data_files/MathJax.js"></script>


  <style type="text/css">.MathJax_Hover_Frame {border-radius: .25em; -webkit-border-radius: .25em; -moz-border-radius: .25em; -khtml-border-radius: .25em; box-shadow: 0px 0px 15px #83A; -webkit-box-shadow: 0px 0px 15px #83A; -moz-box-shadow: 0px 0px 15px #83A; -khtml-box-shadow: 0px 0px 15px #83A; border: 1px solid #A6D ! important; display: inline-block; position: absolute}
.MathJax_Hover_Arrow {position: absolute; width: 15px; height: 11px; cursor: pointer}
</style><style type="text/css">#MathJax_About {position: fixed; left: 50%; width: auto; text-align: center; border: 3px outset; padding: 1em 2em; background-color: #DDDDDD; color: black; cursor: default; font-family: message-box; font-size: 120%; font-style: normal; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; z-index: 201; border-radius: 15px; -webkit-border-radius: 15px; -moz-border-radius: 15px; -khtml-border-radius: 15px; box-shadow: 0px 10px 20px #808080; -webkit-box-shadow: 0px 10px 20px #808080; -moz-box-shadow: 0px 10px 20px #808080; -khtml-box-shadow: 0px 10px 20px #808080; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
.MathJax_Menu {position: absolute; background-color: white; color: black; width: auto; padding: 5px 0px; border: 1px solid #CCCCCC; margin: 0; cursor: default; font: menu; text-align: left; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; z-index: 201; border-radius: 5px; -webkit-border-radius: 5px; -moz-border-radius: 5px; -khtml-border-radius: 5px; box-shadow: 0px 10px 20px #808080; -webkit-box-shadow: 0px 10px 20px #808080; -moz-box-shadow: 0px 10px 20px #808080; -khtml-box-shadow: 0px 10px 20px #808080; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
.MathJax_MenuItem {padding: 1px 2em; background: transparent}
.MathJax_MenuArrow {position: absolute; right: .5em; color: #666666}
.MathJax_MenuActive .MathJax_MenuArrow {color: white}
.MathJax_MenuArrow.RTL {left: .5em; right: auto}
.MathJax_MenuCheck {position: absolute; left: .7em}
.MathJax_MenuCheck.RTL {right: .7em; left: auto}
.MathJax_MenuRadioCheck {position: absolute; left: .7em}
.MathJax_MenuRadioCheck.RTL {right: .7em; left: auto}
.MathJax_MenuLabel {padding: 1px 2em 3px 1.33em; font-style: italic}
.MathJax_MenuRule {border-top: 1px solid #DDDDDD; margin: 4px 3px}
.MathJax_MenuDisabled {color: GrayText}
.MathJax_MenuActive {background-color: #606872; color: white}
.MathJax_Menu_Close {position: absolute; width: 31px; height: 31px; top: -15px; left: -15px}
</style><style type="text/css">.MathJax_Preview .MJXc-math {color: inherit!important}
</style><style type="text/css">#MathJax_Zoom {position: absolute; background-color: #F0F0F0; overflow: auto; display: block; z-index: 301; padding: .5em; border: 1px solid black; margin: 0; font-weight: normal; font-style: normal; text-align: left; text-indent: 0; text-transform: none; line-height: normal; letter-spacing: normal; word-spacing: normal; word-wrap: normal; white-space: nowrap; float: none; box-shadow: 5px 5px 15px #AAAAAA; -webkit-box-shadow: 5px 5px 15px #AAAAAA; -moz-box-shadow: 5px 5px 15px #AAAAAA; -khtml-box-shadow: 5px 5px 15px #AAAAAA; filter: progid:DXImageTransform.Microsoft.dropshadow(OffX=2, OffY=2, Color='gray', Positive='true')}
#MathJax_ZoomOverlay {position: absolute; left: 0; top: 0; z-index: 300; display: inline-block; width: 100%; height: 100%; border: 0; padding: 0; margin: 0; background-color: white; opacity: 0; filter: alpha(opacity=0)}
#MathJax_ZoomFrame {position: relative; display: inline-block; height: 0; width: 0}
#MathJax_ZoomEventTrap {position: absolute; left: 0; top: 0; z-index: 302; display: inline-block; border: 0; padding: 0; margin: 0; background-color: white; opacity: 0; filter: alpha(opacity=0)}
</style><style type="text/css">.MathJax_Preview {color: #888}
#MathJax_Message {position: fixed; left: 1em; bottom: 1.5em; background-color: #E6E6E6; border: 1px solid #959595; margin: 0px; padding: 2px 8px; z-index: 102; color: black; font-size: 80%; width: auto; white-space: nowrap}
#MathJax_MSIE_Frame {position: absolute; top: 0; left: 0; width: 0px; z-index: 101; border: 0px; margin: 0px; padding: 0px}
.MathJax_Error {color: #CC0000; font-style: italic}
</style><style type="text/css">.MJXc-script {font-size: .8em}
.MJXc-right {-webkit-transform-origin: right; -moz-transform-origin: right; -ms-transform-origin: right; -o-transform-origin: right; transform-origin: right}
.MJXc-bold {font-weight: bold}
.MJXc-italic {font-style: italic}
.MJXc-scr {font-family: MathJax_Script,'Times New Roman',Times,STIXGeneral,serif}
.MJXc-frak {font-family: MathJax_Fraktur,'Times New Roman',Times,STIXGeneral,serif}
.MJXc-sf {font-family: MathJax_SansSerif,'Times New Roman',Times,STIXGeneral,serif}
.MJXc-cal {font-family: MathJax_Caligraphic,'Times New Roman',Times,STIXGeneral,serif}
.MJXc-mono {font-family: MathJax_Typewriter,'Times New Roman',Times,STIXGeneral,serif}
.MJXc-largeop {font-size: 150%}
.MJXc-largeop.MJXc-int {vertical-align: -.2em}
.MJXc-math {display: inline-block; line-height: 1.2; text-indent: 0; font-family: 'Times New Roman',Times,STIXGeneral,serif; white-space: nowrap; border-collapse: collapse}
.MJXc-display {display: block; text-align: center; margin: 1em 0}
.MJXc-math span {display: inline-block}
.MJXc-box {display: block!important; text-align: center}
.MJXc-box:after {content: " "}
.MJXc-rule {display: block!important; margin-top: .1em}
.MJXc-char {display: block!important}
.MJXc-mo {margin: 0 .15em}
.MJXc-mfrac {margin: 0 .125em; vertical-align: .25em}
.MJXc-denom {display: inline-table!important; width: 100%}
.MJXc-denom > * {display: table-row!important}
.MJXc-surd {vertical-align: top}
.MJXc-surd > * {display: block!important}
.MJXc-script-box > *  {display: table!important; height: 50%}
.MJXc-script-box > * > * {display: table-cell!important; vertical-align: top}
.MJXc-script-box > *:last-child > * {vertical-align: bottom}
.MJXc-script-box > * > * > * {display: block!important}
.MJXc-mphantom {visibility: hidden}
.MJXc-munderover {display: inline-table!important}
.MJXc-over {display: inline-block!important; text-align: center}
.MJXc-over > * {display: block!important}
.MJXc-munderover > * {display: table-row!important}
.MJXc-mtable {vertical-align: .25em; margin: 0 .125em}
.MJXc-mtable > * {display: inline-table!important; vertical-align: middle}
.MJXc-mtr {display: table-row!important}
.MJXc-mtd {display: table-cell!important; text-align: center; padding: .5em 0 0 .5em}
.MJXc-mtr > .MJXc-mtd:first-child {padding-left: 0}
.MJXc-mtr:first-child > .MJXc-mtd {padding-top: 0}
.MJXc-mlabeledtr {display: table-row!important}
.MJXc-mlabeledtr > .MJXc-mtd:first-child {padding-left: 0}
.MJXc-mlabeledtr:first-child > .MJXc-mtd {padding-top: 0}
.MJXc-merror {background-color: #FFFF88; color: #CC0000; border: 1px solid #CC0000; padding: 1px 3px; font-style: normal; font-size: 90%}
.MJXc-scale0 {-webkit-transform: scaleX(.0); -moz-transform: scaleX(.0); -ms-transform: scaleX(.0); -o-transform: scaleX(.0); transform: scaleX(.0)}
.MJXc-scale1 {-webkit-transform: scaleX(.1); -moz-transform: scaleX(.1); -ms-transform: scaleX(.1); -o-transform: scaleX(.1); transform: scaleX(.1)}
.MJXc-scale2 {-webkit-transform: scaleX(.2); -moz-transform: scaleX(.2); -ms-transform: scaleX(.2); -o-transform: scaleX(.2); transform: scaleX(.2)}
.MJXc-scale3 {-webkit-transform: scaleX(.3); -moz-transform: scaleX(.3); -ms-transform: scaleX(.3); -o-transform: scaleX(.3); transform: scaleX(.3)}
.MJXc-scale4 {-webkit-transform: scaleX(.4); -moz-transform: scaleX(.4); -ms-transform: scaleX(.4); -o-transform: scaleX(.4); transform: scaleX(.4)}
.MJXc-scale5 {-webkit-transform: scaleX(.5); -moz-transform: scaleX(.5); -ms-transform: scaleX(.5); -o-transform: scaleX(.5); transform: scaleX(.5)}
.MJXc-scale6 {-webkit-transform: scaleX(.6); -moz-transform: scaleX(.6); -ms-transform: scaleX(.6); -o-transform: scaleX(.6); transform: scaleX(.6)}
.MJXc-scale7 {-webkit-transform: scaleX(.7); -moz-transform: scaleX(.7); -ms-transform: scaleX(.7); -o-transform: scaleX(.7); transform: scaleX(.7)}
.MJXc-scale8 {-webkit-transform: scaleX(.8); -moz-transform: scaleX(.8); -ms-transform: scaleX(.8); -o-transform: scaleX(.8); transform: scaleX(.8)}
.MJXc-scale9 {-webkit-transform: scaleX(.9); -moz-transform: scaleX(.9); -ms-transform: scaleX(.9); -o-transform: scaleX(.9); transform: scaleX(.9)}
.MathJax_CHTML .noError {vertical-align: ; font-size: 90%; text-align: left; color: black; padding: 1px 3px; border: 1px solid}
</style></head>

  <body><div id="MathJax_Message" style="display: none;"></div>
    <div class="navbar">
      <div class="navbar-inner">
        <div class="container-narrow">
          <a class="brand" href="http://genomicsclass.github.io/book">PH525x series - Biomedical Data Science</a>
        </div>
      </div>
    </div>

    <div class="container-narrow">

      <div class="content">
        
<div class="page-header">
  <h2>Basic inference for high-throughput data </h2>
</div>

<div class="row-fluid">
  <div class="span12">
    <pre><code class="language-r"><span class="keyword">library</span><span class="paren">(</span><span class="identifier">rafalib</span><span class="paren">)</span>
</code></pre>

<h1 id="introduction">Introduction</h1>

<p>Supposed we were given high-throughput gene expression data that was measured for several individuals in two populations. We are asked to report which genes have different average expression levels in the two populations. Note that if, instead thousands of genes, we were handed data from just one gene we could simply apply  the inference techniques that we have learned before. We could, for example, use a t-test or some other test. Here we review what changes when we consider high-throughput data.</p>

<h1 id="thousands-of-test">Thousands of test</h1>

<p>In this data we have two groups denoted with 0 and 1:</p>

<pre><code class="language-r"><span class="keyword">library</span><span class="paren">(</span><span class="identifier">GSE5859Subset</span><span class="paren">)</span>
<span class="identifier">data</span><span class="paren">(</span><span class="identifier">GSE5859Subset</span><span class="paren">)</span>
<span class="identifier">g</span> <span class="operator">&lt;-</span> <span class="identifier">sampleInfo</span><span class="operator">$</span><span class="identifier">group</span>
<span class="identifier">g</span>
</code></pre>

<pre><code>##  [1] 1 1 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0 0 0
</code></pre>

<p>If we were interested in a particular gene, let’s arbitrarily pick the one on the 25th row, we would simply compute a t-test; assuming the data is well approximated by normal:</p>

<pre><code class="language-r"><span class="identifier">e</span> <span class="operator">&lt;-</span> <span class="identifier">geneExpression</span><span class="paren">[</span><span class="number">25</span>,<span class="paren">]</span>
<span class="identifier">mypar2</span><span class="paren">(</span><span class="number">1</span>,<span class="number">2</span><span class="paren">)</span>
<span class="identifier">qqnorm</span><span class="paren">(</span><span class="identifier">e</span><span class="paren">[</span><span class="identifier">g</span><span class="operator">==</span><span class="number">1</span><span class="paren">]</span><span class="paren">)</span>
<span class="identifier">qqline</span><span class="paren">(</span><span class="identifier">e</span><span class="paren">[</span><span class="identifier">g</span><span class="operator">==</span><span class="number">1</span><span class="paren">]</span><span class="paren">)</span>
<span class="identifier">qqnorm</span><span class="paren">(</span><span class="identifier">e</span><span class="paren">[</span><span class="identifier">g</span><span class="operator">==</span><span class="number">0</span><span class="paren">]</span><span class="paren">)</span>
<span class="identifier">qqline</span><span class="paren">(</span><span class="identifier">e</span><span class="paren">[</span><span class="identifier">g</span><span class="operator">==</span><span class="number">0</span><span class="paren">]</span><span class="paren">)</span>
</code></pre>

<p><img src="./Basic inference for high-throughput data_files/inference_for_highthroughput-unnamed-chunk-3-1.png" alt="plot of chunk unnamed-chunk-3"> </p>

<p>The qq-plots show that the data is well approximated by the normal approximation so apply a t-test. The t-test does not find this gene to be statistically significant:</p>

<pre><code class="language-r"><span class="identifier">t.test</span><span class="paren">(</span><span class="identifier">e</span><span class="paren">[</span><span class="identifier">g</span><span class="operator">==</span><span class="number">1</span><span class="paren">]</span>,<span class="identifier">e</span><span class="paren">[</span><span class="identifier">g</span><span class="operator">==</span><span class="number">0</span><span class="paren">]</span><span class="paren">)</span>
</code></pre>

<pre><code>## 
## 	Welch Two Sample t-test
## 
## data:  e[g == 1] and e[g == 0]
## t = 0.2838, df = 21.217, p-value = 0.7793
## alternative hypothesis: true difference in means is not equal to 0
## 95 percent confidence interval:
##  -0.1431452  0.1884244
## sample estimates:
## mean of x mean of y 
##  10.52505  10.50241
</code></pre>

<p>To answer the question for each gene we simply do this for every gene. Here we will define our own function and use <code>apply</code>:</p>

<pre><code class="language-r"><span class="identifier">myttest</span> <span class="operator">&lt;-</span> <span class="keyword">function</span><span class="paren">(</span><span class="identifier">x</span><span class="paren">)</span> <span class="identifier">t.test</span><span class="paren">(</span><span class="identifier">x</span><span class="paren">[</span><span class="identifier">g</span><span class="operator">==</span><span class="number">1</span><span class="paren">]</span>,<span class="identifier">x</span><span class="paren">[</span><span class="identifier">g</span><span class="operator">==</span><span class="number">0</span><span class="paren">]</span>,<span class="identifier">var.equal</span><span class="operator">=</span><span class="literal">TRUE</span><span class="paren">)</span><span class="operator">$</span><span class="identifier">p.value</span>
<span class="identifier">pvals</span> <span class="operator">&lt;-</span> <span class="identifier">apply</span><span class="paren">(</span><span class="identifier">geneExpression</span>,<span class="number">1</span>,<span class="identifier">myttest</span><span class="paren">)</span>
</code></pre>

<p>We can now see which genes have p-values less than, say, 0.05. For example right away we see that:</p>

<pre><code class="language-r"><span class="identifier">sum</span><span class="paren">(</span><span class="identifier">pvals</span><span class="operator">&lt;</span><span class="number">0.05</span><span class="paren">)</span>
</code></pre>

<pre><code>## [1] 1383
</code></pre>

<p>genes had p-values less than 0.05</p>

<p>However, as we will describe in more detail below, we have to be careful in interpreting this result because we have performed over  8,000 test. Note that if we performed the same procedure on random data, for which the null hypothesis is true for all feature, we obtain the following results:</p>

<pre><code class="language-r"><span class="identifier">set.seed</span><span class="paren">(</span><span class="number">1</span><span class="paren">)</span>
<span class="identifier">m</span> <span class="operator">&lt;-</span> <span class="identifier">nrow</span><span class="paren">(</span><span class="identifier">geneExpression</span><span class="paren">)</span>
<span class="identifier">n</span> <span class="operator">&lt;-</span> <span class="identifier">ncol</span><span class="paren">(</span><span class="identifier">geneExpression</span><span class="paren">)</span>
<span class="identifier">randomData</span> <span class="operator">&lt;-</span> <span class="identifier">matrix</span><span class="paren">(</span><span class="identifier">rnorm</span><span class="paren">(</span><span class="identifier">n</span><span class="operator">*</span><span class="identifier">m</span><span class="paren">)</span>,<span class="identifier">m</span>,<span class="identifier">n</span><span class="paren">)</span>
<span class="identifier">nullpvals</span> <span class="operator">&lt;-</span> <span class="identifier">apply</span><span class="paren">(</span><span class="identifier">randomData</span>,<span class="number">1</span>,<span class="identifier">myttest</span><span class="paren">)</span>
<span class="identifier">sum</span><span class="paren">(</span><span class="identifier">nullpvals</span><span class="operator">&lt;</span><span class="number">0.05</span><span class="paren">)</span>
</code></pre>

<pre><code>## [1] 419
</code></pre>

<p>As we will explain in the chapter, this is to be expected. Note that 419 is roughly 0.05*8192 and we will describe the theory that tells us why this prediction works.</p>

<h1 id="faster-implementation-of-t-test">Faster implementation of t-test</h1>

<p>Before, we continue, we should note that the above implementation is very inefficient. There are several faster implementations that perform t-test for high throughput data. For example</p>

<pre><code class="language-r"><span class="keyword">library</span><span class="paren">(</span><span class="identifier">genefilter</span><span class="paren">)</span>
<span class="identifier">results</span> <span class="operator">&lt;-</span> <span class="identifier">rowttests</span><span class="paren">(</span><span class="identifier">geneExpression</span>,<span class="identifier">factor</span><span class="paren">(</span><span class="identifier">g</span><span class="paren">)</span><span class="paren">)</span>
<span class="identifier">max</span><span class="paren">(</span><span class="identifier">abs</span><span class="paren">(</span><span class="identifier">pvals</span><span class="operator">-</span><span class="identifier">results</span><span class="operator">$</span><span class="identifier">p</span><span class="paren">)</span><span class="paren">)</span>
</code></pre>

<pre><code>## [1] 6.528111e-14
</code></pre>

<p>Note that <code>genefilter</code> is available from the Bioconductor projects. In a later section we will say more about this project but here is how one installs it:</p>

<pre><code class="language-r"><span class="keyword">source</span><span class="paren">(</span><span class="string">"http://www.bioconductor.org/biocLite.R"</span><span class="paren">)</span>
<span class="identifier">biocLite</span><span class="paren">(</span><span class="string">"genefilter"</span><span class="paren">)</span>
</code></pre>

<p>Note that we get practically the same answer and much faster performance.</p>


  </div>
</div>


      </div>
      <hr>
      <footer>
        <p><small>
  <!-- start of Karl's footer; modify this part -->

<a href="http://genomicsclass.github.io/book/">PH525x</a>, 
Rafael Irizarry and Michael Love, 
<a href="https://github.com/genomicsclass/labs/blob/master/LICENSE">MIT License</a>

  <!-- end of Karl's footer; modify this part -->
        </small></p>
      </footer>

    </div>

    
  


</body></html>